### DISCUSSION TITLE
Some how stitching two data sources together under one entry in Gatsby's graphql schema

### DISCUSSION THREAD

**Discussion by wildpow** [gatsbyjs/gatsby#28275] [Help] [answered]
My company uses two sources for data. DatoCMS for page/product content (using `gatsby-source-datocms`) and Shopify for product pricing/inventory (using `gatsby-source-Shopify`). Product content is connected to product pricing through a `Shopify-id` plug-in on DatoCMS. This has been fine for single product pages. In the `gatsby-node.js`, we use `createPages`, pass the Shopify ids through context to the template, and the template queries the Shopify data. 

The problems start when we need that same pricing in pages that are not programmatically created, like a page that listed all products. I run a page query to get the content and get the Shopify ids but, now I'm stuck.

My first thought was to pass the ids to the static query component, but this feature will never happen from what I have read. My next idea was to stitch these two data sources together in Gatsby's schema into one entry or create a field on the DatCMS nodes and inject the Shopify data. I've never dug into Gatsby's internal APIs or seen an example of this to know if it's even possible. Am I going down a fruitless rabbit hole, or is there another solution I'm not seeing?

**Comment by vladar** [comment]
@wildpow You can absolutely stitch types together. But that will require diving deep into schema customization API. See this blog post: https://www.gatsbyjs.com/blog/2019-03-18-releasing-new-schema-customization/

Maybe this example will be a good start:

https://github.com/gatsbyjs/gatsby/blob/5bc84a9fa3bf4058aa7b5df8e47a8819a0d8392f/examples/using-type-definitions/gatsby-node.js#L28-L40

**Reply by wildpow** [reply]
I'm having trouble wrapping my head around taking a list of products for one source and combining the data from another source based on a shared field `shopifyId` === `shopifyConnection`.
My solution since I made my original post has been to use `useStaticQuery` but, doing this for every page that needs this combination of data seems unmanageable.

```javascript
export default () => {
  const { shopifyProducts, datoProduct } = useStaticQuery(
    graphql`
      query products {
        shopifyProducts: allShopifyProduct {
          nodes {
            shopifyId
            priceRange {
              minVariantPrice {
                amount
              }
              maxVariantPrice {
                amount
              }
            }
          }
        }
        datoProducts: allDatoCmsProduct {
            nodes {
              shopifyConnection
              slug
              id
              name
              saleInfo {
                saleBanner
              }
            }
          }
      }
    `,
  );
  const combinedProduct = [];
  shopifyProducts.nodes.forEach((shopifyItem) => {
    datoProducts.nodes.forEach((datoItem) => {
      if (datoItem.shopifyConnection === shopifyItem.shopifyId) {
        combinedProduct.push({ ...datoItem, ...shopifyItem.priceRange });
      }
    });
  });
  return { products: combinedProduct };
};
```

**Reply by wildpow** [reply]
My other thought was to use `onCreatePage` and pass combined data through `pageContext` but, it seems you can't get `graphql` inside the function to make the initial queries.

**Reply by vladar** [reply]
Have you tried the schema customization approach I've mentioned above? It will be something like this:

```js
// in gatsby-node.js

exports.createSchemaCustomization = ({ actions, schema }) => {
  const ExtendedDatoCmsProduct = schema.buildObjectType({
    name: `DatoCmsProduct`,
    fields: {
      shopifyPriceRange: {
        type: `ShopifyProductPriceRange`,
        resolve(root, args, context, info) {
          const node = context.nodeModel.runQuery({
            query: { filter: { shopifyId: { eq: root.shopifyConnection } } },
            type: `ShopifyProduct`,
            firstOnly: true,
          })
          return node ? node.priceRange : null
        }
      }
    }
  })

  actions.createTypes([
    ExtendedDatoCmsProduct 
  ])
}
```

Writing blindly so there might be errors there but it should give you some direction. Also, check the blog post and the example above.

**Reply by wildpow** [reply]
With a little modification of your example, I was able to get a new field of `shopifyPriceRange` with a value of `null`, but both `root.shopifyConnection` and `node.priceRange` were coming up as `undefined`.

My first go at it using the docs and your example I came up with this:
```javascript
exports.createSchemaCustomization = ({ actions, schema }) => {
  const ExtendedDatoCmsProuct = schema.buildObjectType({
    name: `DatoCmsMattress`,
    fields: {
      shopifyPriceRange: {
        type: `ShopifyProductPriceRange`,
        resolve(root, args, context, info) {
          const node = context.nodeModel.getAllNodes({
            type: `ShopifyProduct`,
          });
          return node.map((item) =>
            item.shopifyId ===
            root.entityPayload.attributes.shop_matt_connection
              ? item.priceRange
              : null,
          );
        },
      },
    },
  });
  actions.createTypes([ExtendedDatoCmsProuct]);
};
```
GraphiQL results: 
```
 {
    "shopifyPriceRange": {
          "minVariantPrice": null,
          "maxVariantPrice": null
       }
 },
```
Which seem to be getting closer, still no data but, I'll keep messing with it. 

Thanks for your help so far.

**Reply by wildpow** [reply]
Using either your example or the one from my last post. I get the new field but null for the data. Also, trying to console log the results of `runQuery` like in your example, I get back `Promise { <pending> }`. Is that what I should expect?

**Comment by wildpow** [comment]
Here is my only success so far:
```javascript
exports.createResolvers = ({ createResolvers }) => {
  createResolvers({
    DatoCmsMattress: {
      shopifyInfo: {
        type: [`ShopifyProduct`],
        resolve(source, args, context, info) {
          const fieldValue =
            source.entityPayload.attributes.shop_matt_connection;
          const shopifyMatts = context.nodeModel.getAllNodes({
            type: "ShopifyProduct",
          });
          return shopifyMatts.filter((matt) => matt.shopifyId === fieldValue);
        },
      },
    },
  });
};
```
Which adds the entire Shopify product to Dato's entry. Not exactly what I was looking for, but it works. I got the idea from the same file you posted.
https://github.com/gatsbyjs/gatsby/blob/5bc84a9fa3bf4058aa7b5df8e47a8819a0d8392f/examples/using-type-definitions/gatsby-node.js#L88-L106

Since my development build times are so long anyway, I'm not sure I could measure a negative impact if there were one.

**Reply by vladar** [reply]
I discourage you from using `getAllNodes` and looping through all nodes as it is expensive and will be pretty slow with many nodes. Have you tried using `runQuery` instead? It uses indexing under the hood to find the required data:

```diff
-          const shopifyMatts = context.nodeModel.getAllNodes({
-            type: "ShopifyProduct",
-          });
-          return shopifyMatts.filter((matt) => matt.shopifyId === fieldValue);
+          return context.nodeModel.runQuery({
+            query: { filter: { shopifyId: { eq: fieldValue } } },
+            type: `ShopifyProduct`,
+            firstOnly: true,
+          })
```

**Reply by wildpow** [reply]
This is what I get back in GraphiQL:

![Screen Shot 2020-12-08 at 10 38 09 AM](https://user-images.githubusercontent.com/17464319/101526492-8a3f6500-3941-11eb-8748-17f986df4e3c.png)

**Comment by wildpow** [comment, marked as answer]
@vladar This works:

```javascript
exports.createResolvers = ({ createResolvers }) => {
  createResolvers({
    DatoCmsMattress: {
      shopifyInfo: {
        type: [`ShopifyProduct`],
        resolve(source, args, context, info) {
          const fieldValue =
            source.entityPayload.attributes.shop_matt_connection;
          return context.nodeModel.runQuery({
            query: { filter: { shopifyId: { eq: fieldValue } } },
            type: `ShopifyProduct`,
            // Not sure why removing this made it work?!?
            // firstOnly: true,
          });
        },
      },
    },
  });
};

```

**Reply by vladar** [reply]
Oh, yeah, I've missed that you expect to get an array of values. for this field. `firstOnly` simply extracts the first value from this array and returns it. So it would've worked if your type was `type: "ShopifyProduct"`. But for `type: ["ShopifyProduct"]` the `firstOnly` flag is not needed.
